#' Return line plots of cluster metadata
#'
#' `fa_clusterDiversity_metaplot` returns 3x subplots that show the unique sequence count, total read count, and average LED per cluster.
#'
#' @param diversityDF A data.frame generated by `fa_clusterDiversity`
#' @param xaxis The name assigned to all x-axes in the plot; this axis corresponds to the cluster number (default: "Cluster")
#' @param yaxes The 3x y-axis names respectively assigned to the 3x subplots; these respectively correspond to number of unique sequences, number of total reads, and average LED (defaults: "Seq. count", "Read count", "Avg. LED")
#' @param plot_title The title of the plot (default: "Cluster metaplots")
#' @param line_colours The 3x colors respectively used for the 3x subplots (defaults: "blue", "orange", "forestgreen")
#'
#' @seealso [grDevices::colors()], [fa_clusterDiversity()], [fa_clusterLED()], [fa_clusterDiversity_kmer()]
#'
#' @return This function returns 3x interactive plots of cluster metastats
#' @export
#'
#' @examples
#' \dontrun{
#' # count FASTQ file, cluster data, compute metadata
#' countData <- fa_count(dataInput = "PATH/TO/FILE.fastq")
#' clusterData <- fa_clusterLED(faDF = countData)
#' diversityDF <- fa_clusterDiversity(faDF_cluster = clusterData)
#'
#' # plot metadata
#' fa_clusterDiversity_metaplot(diversityDF = diversityDF)
#' }
#'
#' @importFrom dplyr %>%
#' @importFrom rlang .data
fa_clusterDiversity_metaplot <- function(
    diversityDF = NULL,
    xaxis = "Cluster",
    yaxes = c("Seq. count", "Read count", "Avg. LED"),
    plot_title = "Cluster metaplots",
    line_colours = c("blue", "orange", "forestgreen")
){

  # line plot of unique sequences per cluster
  seqsPlot <- boldPlots(
    ggplot2::ggplot(diversityDF, ggplot2::aes(x = .data$Cluster, y = .data$TotalSequences)) +
      ggplot2::geom_line(colour = line_colours[1], linewidth = 2) +
      ggplot2::labs(x = xaxis, y = yaxes[1]) +
      ggplot2::theme_bw()
  ) %>% plotly::ggplotly()

  # line plot of total sequences per cluster
  readsPlot <- boldPlots(
    ggplot2::ggplot(diversityDF, ggplot2::aes(x = .data$Cluster, y = .data$TotalReads)) +
      ggplot2::geom_line(colour = line_colours[2], linewidth = 2) +
      ggplot2::labs(x = xaxis, y = yaxes[2]) +
      ggplot2::theme_bw()
  ) %>% plotly::ggplotly()

  # line plot of average LED to seed per cluster
  ledPlot <- boldPlots(
    ggplot2::ggplot(diversityDF, ggplot2::aes(x = .data$Cluster, y = .data$AverageLED)) +
      ggplot2::geom_line(colour = line_colours[3], linewidth = 2) +
      ggplot2::labs(x = xaxis, y = yaxes[3]) +
      ggplot2::theme_bw()
  ) %>% plotly::ggplotly()

  # make interactive figure from subplots
  fig <- plotly::subplot(seqsPlot, readsPlot, ledPlot, titleY = TRUE, shareX = TRUE, nrows = 3) %>%
    plotly::layout(title = paste0("<b>", plot_title, "</b>"), margin = list(b = 50, l = 50, r = 50, t = 50)) %>%
    plotly::config(toImageButtonOptions = list(format = "svg", filename = "cluster_metaplots"))

  # return interactive figure
  return(fig)
}

#' Return a PCA or UMAP scatter plot from a k-mer matrix
#'
#' `fa_clusterDiversity_kmer` returns a two-dimensional PCA or UMAP plot where each point corresponds to one sequence, and colors correspond to user-specified clusters (max = 8 clusters).
#'
#' The function filters the input data.frame for 1-8 user-selected clusters and computes k-mer counts for each sequence
#' The final matrix has one row per sequence and one column per k-mer (e.g., AAA, AAC, etc. in the case of 3-mers), and this is the input to a PCA or UMAP.
#' Only the first two dimensions of PCA / UMAP are plotted, and points have colors corresponding to their respective clusters.
#'
#' @param faDF_cluster A clustered data.frame generated by fa_clusterLED
#' @param kmerSize Value of k used to generate k-mers (e.g., 3 generates 3-mers)
#' @param clustersToPlot Clusters to include in the visualization; clusters are named as integers
#' @param plotType One of `PCA` or `UMAP` (default: "PCA")
#' @param xaxis Name for the x-axis (default: "Dim1")
#' @param yaxis Name for the y-axis (default: "Dim2")
#' @param legend_title Title for the legend (default: "Cluster")
#' @param plot_title Title for the plot (default: "Cluster k-mer plot")
#' @param colour_palette_disc A discrete RColorBrewer palette (default: "Dark2")
#'
#' @seealso [RColorBrewer::brewer.pal.info], [fa_clusterDiversity()], [fa_clusterLED()], [fa_clusterDiversity_metaplot()]
#'
#' @return This function returns a PCA or UMAP plot based on kmer counts
#' @export
#'
#' @examples
#' \dontrun{
#' # count FASTQ file, cluster data
#' countData <- fa_count(dataInput = "PATH/TO/FILE.fastq")
#' clusterData <- fa_clusterLED(faDF = countData)
#'
#' fa_clusterDiversity_kmer(faDF_cluster = clusterData, kmerSize = 3, clustersToPlot = c(1:5))
#' }
#'
#' @importFrom dplyr %>%
#' @importFrom rlang .data
fa_clusterDiversity_kmer <- function(
    faDF_cluster = NULL,
    kmerSize = 3,
    clustersToPlot = c(1:3),
    plotType = "PCA",
    xaxis = "Dim1",
    yaxis = "Dim2",
    legend_title = "Cluster",
    plot_title = "Cluster k-mer plot",
    colour_palette_disc = "Dark2"
){

  # filter for user-defined clusters
  faDF_cluster <- faDF_cluster %>% dplyr::filter(.data$Cluster %in% clustersToPlot)

  # compute kmer matrix
  kmerMatrix <- faDF_cluster$Sequences %>%
    as.character() %>%
    strsplit(split = "") %>%
    kmer::kcount(x = .data, k = kmerSize, named = FALSE)
  rownames(kmerMatrix) <- NULL

  # optionally process kmers with PCA
  if(plotType == "PCA"){

    # PCA with kmer matrix; only use 2x PCs
    kmer_coords <- stats::prcomp(kmerMatrix, scale. = FALSE)

    # make plot
    kmer_plot <- factoextra::fviz_pca_ind(
      kmer_coords,
      addEllipses = FALSE, geom = c("point"),
      col.ind = as.factor(faDF_cluster$Cluster), legend.title = legend_title
    ) +
      ggplot2::scale_colour_brewer(palette = colour_palette_disc) +
      ggplot2::labs(x = xaxis, y = yaxis, title = plot_title) +
      ggplot2::theme_bw()
  }

  # optionally process kmers with UMAP
  if(plotType == "UMAP"){

    # UMAP with kmer matrix; only use 2x dimensions
    kmer_coords <- cbind(faDF_cluster, data.frame(umap::umap(kmerMatrix, random_state = 1234)$layout))

    # make plot
    kmer_plot <- ggplot2::ggplot(
      kmer_coords,
      ggplot2::aes(
        .data$X1,
        .data$X2,
        colour = as.factor(.data$Cluster),
        text = glue::glue("UMAP1: {round(.data$X1, 3)}, UMAP2: {round(.data$X2, 3)}, Cluster: {as.factor(.data$Cluster)}")
      )
    ) +
      ggplot2::geom_point() +
      ggplot2::scale_colour_brewer(palette = colour_palette_disc) +
      ggplot2::labs(x = xaxis, y = yaxis, title = plot_title, colour = legend_title) +
      ggplot2::theme_bw()
  }

  # make bold text
  kmer_plot <- boldPlots(p = kmer_plot)

  # make interactive figure
  fig <- plotly::ggplotly(kmer_plot, tooltip = "text") %>%
    plotly::config(toImageButtonOptions = list(format = "svg", filename = "kmer_cluster"))

  # return interactive PCA plot
  return(fig)
}
